shards: introduce a scheduler abstraction

keegancsmith opened the pull request:
This PR introduces a scheduler abstraction for searches to use. Every search now starts in the "interactive" queue. But after 5s will move to the "batch" queue. The idea here is to have simple approach of allowing long running queries to run without impacting interactive queries.

We use WeightedSemaphore as the "queue" (internally has a FIFO queue). Before this PR we just had one WeightedSemaphore with size #CPUs. In this PR that semaphore becomes the "interactive" queue. The "batch" queue is a new semaphore with size 1/4 #CPUs. This doesn't limit the amount of CPU a search can consume, but rather we use number of CPUs as a guide for the number of concurrent searches that can be running.

This commit changes a few things, so is intended to be reviewed commit by commit.

What is still missing is some observability. That will be done as part of this PR so we can monitor the effects.

Part of https://github.com/sourcegraph/sourcegraph/issues/18305

keegancsmith committed:
shards: introduce scheduler abstraction over throttle

We want to make our throttling take into account longer running
searches. As such we introduce a lightweight scheduler and process
type. In this commit the scheduler is behaving exactly the same as the
pre-existing semaphore based throttler. We introduce this in a separate
commit to make the functional change in the next commit clearer.

Change-Id: I4e2d8a11a3c72f76fef18fa623958845d9d2fcd8

keegancsmith committed:
shards: introduce process.Yield

The actual yield logic is not implemented yet, but this commit adds the
necessary scaffolding to make the scheduler yielding logic self
contained in the next commit.

We document what we expect the scheduler to acheive with
yield. We only want to call yield once (when we downgrade a search), so
we introduce a deadlineTimer to help track when that should occur. We
add some benchmarks to ensure this is reasonably efficient:

  goos: darwin
  goarch: amd64
  pkg: github.com/google/zoekt/shards
  cpu: Intel(R) Core(TM) i7-8559U CPU @ 2.70GHz
  BenchmarkYield/timer-8         	349413274	        3.449 ns/op
  BenchmarkYield/now-8           	17178878	       65.88 ns/op
  BenchmarkYield/deadlineTimer-8 	252081333	        4.712 ns/op
  BenchmarkYield/yield-8         	170765276	        7.012 ns/op

Change-Id: Iabb05825fc0cd63910832a3dab3033e91a9cb773

keegancsmith committed:
shards: use a RWMutex for exclusive vs search processes

When we introduce multiple semaphores for batch vs interactive, we need
a single mutex to lock over all semaphores. This is mu.

Change-Id: Id551bf14121e4e5bebb56de457d2500fd6129902

keegancsmith committed:
shards: implement a batch queue

We now have two semaphores, one for interactive queries and one for
batch. The batch size is 1/4 of interactive. When we downgrade a
process (after 5s) we release the interactive semaphore and acquire the
batch semaphore. We carefully handle the case of acquiring failing
during this process.

The RWMutex is always held until our final call to Release to prevent an
exclusive process acquiring while a process is switching queues.

Change-Id: Ibd0ea99abec809a67fabcb40b410a11ad43d9557

keegancsmith committed:
shards: call yield during Search

To prevent concurrent calls to yield we adjust the feeder of shards to
call yield. This does mean we may do a few more searches while yield is
blocking, but this is acceptable for simplicity.

Change-Id: I75e04042ba8992817b2f6546dff19d3e92580695

stefanhengl reviewed (approved):
I had first walk through the PR. Looks really great! I really like how confined the change is to the shards package. Since it is quite a big change I will let it settle a bit and go over it again later. 

Thanks for the nice separation in commits. Much appreciated!

emidoots commented:
Nifty solution / idea!

emidoots commented:
two thoughst:

* This could make a lot of existing heavy API queries fail. e.g. if they are today close to the 2m search timeout, this could make them 4x slower and result in a "all our API queries are failing after the upgrade" - unsure the best way to fix this, maybe raise the max search timeout to 4x / 8m? 
* This could make site admins who issue really heavy search queries as a means to test search performance before upgrading to a new Sourcegraph version skeptical ("it's 4x slower!") e.g. https://app.hubspot.com/contacts/2762526/company/407948923/ does this - key here would just be to clearly communicate this change and why

stefanhengl commented:
> his could make them 4x slower and result in a "all our API queries are failing after the upgrade"

@slimsag Can you elaborate on this? How are queries getting slower? The two semaphores should offer the same performance, just different levels of concurrency. As long as the customer doesn't issue more heavy queries than there are slots in semBatch then there should be no problem.

I imagine that in the future we let the customer configure the scheduler (size of semBatch v. semInteractive and threshold time).

emidoots commented:
Ahh, that makes sense - what I said was just based on the PR description “after
5s will move to the "batch" queue. The batch queue can ony have 1/4 of CPUs”

Sounds like the implementation is smarter than that :)

On Mon, Mar 15, 2021 at 12:56 AM Stefan Hengl ***@***.***>
wrote:

> his could make them 4x slower and result in a "all our API queries are
> failing after the upgrade"
>
> @slimsag <https://github.com/slimsag> Can you elaborate on this? How are
> queries getting slower? The two semaphores should offer the same
> performance, just different levels of concurrency. As long as the customer
> doesn't issue more heavy queries than there are slots in semBatch then
> there should be no problem.
>
> I imagine that in the future we let the customer configure the scheduler
> (size of semBatch v. semInteractive and threshold time).
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/sourcegraph/zoekt/pull/74#issuecomment-799200323>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAYGWODG62TJ3IGVNVJNLELTDW4THANCNFSM4ZCA47BA>
> .
>
-- 
Follow me on twitter @slimsag <https://twitter.com/slimsag>.

keegancsmith commented:
I see the confusion. Will update the PR description to match exactly what we mean by 1/4 CPUs. Thanks.

keegancsmith reviewed (commented):
Thank you for the review. We chatted over VC and to summarise what needs to be done before landing:

- Lots of observability. We want confidence in how this is behaving.
- Lots of local manual testing, with confirmation based on observability added.
- Some hidden temporary tuneables. The hardcoded values may be inappropriate, so we want to be able to adjust behaviour without changing code. Once we have deployed the code on sourcegraph.com and some large customers and they are happy, we can remove the tuneables.
- An escape hatch to the older scheduler. If this goes badly we want to disable it. I doubt it will go badly, but given how this is designed it should be easy for us to do this.

keegancsmith committed:
minor feedback from code review

- clearer comment about batch capacity
- rename quanta to quantum
- comment about why we yield in feeder

Change-Id: Ic1e85fad55f95b6455bcd62602b52989691cb716

keegancsmith committed:
shards: ZOEKTSCHED tuneable to disable scheduler

We introduce a scheduler interface such that we can change the scheduler
at runtime. Now if we are concerned about the new scheduler we can set
the environment variable "ZOEKTSCHED=disable=1". When set we use the old
method of throttling (a single semaphore).

We introduce ZOEKTSCHED which behaves much like the GODEBUG variable. We
do it in this manor so it is cleaner to introduce more tuneables.

Change-Id: I1301f34d479c49b6d172c6f63a257ba34e0d23cd

keegancsmith committed:
shards: tuneables for batch size and interactive duration

These are the two inputs which most affect how the scheduler behaves. So
we make it tuneables such that we can experiment with them.

Change-Id: I97144dbec891f13c29de92adbd3b39925fa8d54e

keegancsmith committed:
Update shards/sched.go

Co-authored-by: Stefan Hengl <stefan@sourcegraph.com>

keegancsmith committed:
shards: add observability around scheduler

We introduce some wrapper types around the RWMutex and Semaphores used
by multiScheduler. These wrappers add prometheus metrics which allow us
to understand that state of the scheduler via gauges and counters.

Change-Id: I8fd3586660101291eafc71ea528c04bfffa666a1

keegancsmith committed:
Update shards/sched.go

Co-authored-by: Stefan Hengl <stefan@sourcegraph.com>

keegancsmith committed:
Update shards/sched.go

Co-authored-by: Stefan Hengl <stefan@sourcegraph.com>

keegancsmith committed:
default to using the new scheduler

Change-Id: Idc16fb1082144beedd9d81118950e6a56ca0eddb
